tower-filter/src/future.rs

```diff
@@ -1,84 +1,91 @@
 //! Future types
 
 use crate::error::{self, Error};
-use futures::{Async, Future, Poll};
+use futures_core::ready;
+use pin_project::{pin_project, project};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
 use tower_service::Service;
 
 /// Filtered response future
+#[pin_project]
 #[derive(Debug)]
 pub struct ResponseFuture<T, S, Request>
 where
     S: Service<Request>,
 {
+    #[pin]
     /// Response future state
     state: State<Request, S::Future>,
 
+    #[pin]
     /// Predicate future
     check: T,
 
     /// Inner service
     service: S,
 }
 
+#[pin_project]
 #[derive(Debug)]
 enum State<Request, U> {
-    Check(Request),
-    WaitResponse(U),
-    Invalid,
+    Check(Option<Request>),
+    WaitResponse(#[pin] U),
 }
 
-impl<T, S, Request> ResponseFuture<T, S, Request>
+impl<F, T, S, Request> ResponseFuture<F, S, Request>
 where
-    T: Future<Error = Error>,
+    F: Future<Output = Result<T, Error>>,
     S: Service<Request>,
     S::Error: Into<error::Source>,
 {
-    pub(crate) fn new(request: Request, check: T, service: S) -> Self {
+    pub(crate) fn new(request: Request, check: F, service: S) -> Self {
         ResponseFuture {
-            state: State::Check(request),
+            state: State::Check(Some(request)),
             check,
             service,
         }
     }
 }
 
-impl<T, S, Request> Future for ResponseFuture<T, S, Request>
+impl<F, T, S, Request> Future for ResponseFuture<F, S, Request>
 where
-    T: Future<Error = Error>,
+    F: Future<Output = Result<T, Error>>,
     S: Service<Request>,
     S::Error: Into<error::Source>,
 {
-    type Item = S::Response;
-    type Error = Error;
+    type Output = Result<S::Response, Error>;
 
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        use self::State::*;
-        use std::mem;
+    #[project]
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
 
         loop {
-            match mem::replace(&mut self.state, Invalid) {
-                Check(request) => {
+            #[project]
+            match this.state.project() {
+                State::Check(request) => {
+                    let request = request
+                        .take()
+                        .expect("we either give it back or leave State::Check once we take");
+
                     // Poll predicate
-                    match self.check.poll()? {
-                        Async::Ready(_) => {
-                            let response = self.service.call(request);
-                            self.state = WaitResponse(response);
+                    match this.check.as_mut().poll(cx) {
+                        Poll::Ready(r) => {
+                            r?;
```

Comment:
I think ist more preferable here to match also on the result like
```rust
Poll::Ready(Ok(..)) =>
Poll::Ready(Err(e)) =>
```
I personally find this a bit easier to read than using the `r?;`